File: loop_benchmark.go
package main

import (
	"fmt"
	"math/big"
	"os"
	"os/signal"
	"runtime"
	"sync"
	"sync/atomic"
	"syscall"
)

var goVersion string

func main() {
	var totalLoops uint64
	cpuCount := runtime.NumCPU()
	var wg sync.WaitGroup
	wg.Add(cpuCount)

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		<-sigChan
		bigTotalLoops := big.NewInt(0).SetUint64(atomic.LoadUint64(&totalLoops))
		fmt.Printf("%s looped %#v times.\n", goVersion, bigTotalLoops)
		os.Exit(0)
	}()

	for i := 0; i < cpuCount; i++ {
		go func() {
			defer wg.Done()
			var localLoops uint64
			for {
				for j := 0; j < 5000000; j++ {
					// This loop simulates work
				}
				localLoops += 5000000
				atomic.AddUint64(&totalLoops, 5000000)
			}
		}()
	}
	
	wg.Wait()
}

File: loop_benchmark.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <atomic>
#include <csignal>
#include <unistd.h>
#include <iomanip>
#include <sstream>

std::atomic<bool> running(true);
std::atomic<unsigned long long> totalLoops(0);

void signalHandler(int signum) {
    running = false;
}

void loopFunction() {
    unsigned long long localLoops = 0;
    while (running) {
        for (int j = 0; j < 5000000; ++j) {
            // Just a busy loop.
        }
        localLoops += 5000000;
    }
    totalLoops += localLoops;
}

std::string formatNumber(unsigned long long number) {
    std::stringstream ss;
    ss.imbue(std::locale(""));
    ss << std::fixed << number;
    return ss.str();
}

int main() {
    // Register signal handler for SIGINT and SIGTERM
    std::signal(SIGINT, signalHandler);
    std::signal(SIGTERM, signalHandler);

    // Get the number of CPU cores
    unsigned int cpuCount = std::thread::hardware_concurrency();

    // Create a vector to hold the threads
    std::vector<std::thread> threads;

    // Launch a thread for each CPU core
    for (unsigned int i = 0; i < cpuCount; ++i) {
        threads.emplace_back(loopFunction);
    }

    // Wait for all threads to finish
    for (auto& thread : threads) {
        thread.join();
    }

    // Output the result
    std::cout << "C++ looped " << formatNumber(totalLoops) << " times." << std::endl;

    return 0;
}
